evm/p2p/kademlia.py

```diff
@@ -0,0 +1,581 @@
+                return candidates
+            candidates = [c for c in candidates if c not in nodes_seen]
+            logger.info("got {} new candidates".format(len(candidates)))
+            # Add new candidates to nodes_seen so that we don't attempt to bond with failing ones
+            # in the future.
+            nodes_seen.extend(candidates)
+            bonded = yield from asyncio.gather(*[self.bond(c) for c in candidates])
+            logger.info("bonded with {} candidates".format(bonded.count(True)))
+            return [c for c in candidates if bonded[candidates.index(c)]]
+
+        def _exclude_if_asked(nodes):
+            nodes_to_ask = list(set(nodes).difference(nodes_asked))
+            return sort_by_distance(nodes_to_ask, node_id)[:k_find_concurrency]
+
+        closest = self.routing.neighbours(node_id)
+        logger.info("starting lookup; initial neighbours: {}".format(closest))
+        nodes_to_ask = _exclude_if_asked(closest)
+        while nodes_to_ask:
+            logger.info("node lookup; querying {}".format(nodes_to_ask))
+            nodes_asked.extend(nodes_to_ask)
+            results = yield from asyncio.gather(
+                *[_find_node(node_id, n) for n in nodes_to_ask])
+            [closest.extend(candidates) for candidates in results]
+            closest = sort_by_distance(closest, node_id)[:k_bucket_size]
+            nodes_to_ask = _exclude_if_asked(closest)
+
+        logger.info("lookup finished: {}".format(closest))
+        return closest
+
+    # TODO: Run this as a coroutine that loops forever and after each iteration sleeps until the
+    # time when the least recently touched bucket will be considered idle.
+    def refresh_idle_buckets(self):
+        # For buckets that haven't been touched in 3600 seconds, pick a random value in the bucket's
+        # range and perform discovery for that value.
+        for bucket in self.routing.idle_buckets:
+            rid = random.randint(bucket.start, bucket.end)
+            asyncio.ensure_future(self.lookup(rid))
+
+    def _mkpingid(self, echoed, node):
+        pid = force_bytes(echoed) + node.pubkey
+        return pid
+
+    @asyncio.coroutine
+    def populate_not_full_buckets(self):
+        """Go through all buckets that are not full and try to fill them.
+
+        For every node in the replacement cache of every non-full bucket, try to bond.
+        When the bonding succeeds the node is automatically added to the bucket.
+        """
+        for bucket in self.routing.not_full_buckets:
+            for node in bucket.replacement_cache:
+                asyncio.ensure_future(self.bond(node))
+
+
+def sort_by_distance(nodes, target_id):
+    return sorted(nodes, key=operator.methodcaller('id_distance', target_id))
+
+
+def host_port_pubkey_from_uri(uri):
+    node_uri_scheme = b'enode://'
+    pubkey_hex, ip_port = uri[len(node_uri_scheme):].split(b'@')
```

Comment:
we can use `urllib.parse.urlparse` to do this for us.

```python
>>> x = urllib.parse.urlparse('enode://3a514176466fa815ed481ffad09110a2d344f6c9b78c1d14afc351c3a51be33d8072e77939dc03ba44790779b7a1025baf3003f6732430e20cd9b76d953391b3@127.0.0.1:30301')
>>> x.port
30301
>>> x.username
'3a514176466fa815ed481ffad09110a2d344f6c9b78c1d14afc351c3a51be33d8072e77939dc03ba44790779b7a1025baf3003f6732430e20cd9b76d953391b3'
>>> x.hostname
'127.0.0.1'
```

Reply:
Oh, nice!